### PR TITLE
Coastal dataframe assembly from BMI type arrays with example

### PR DESCRIPTION
**Pull request by JurgenZach-NOAA**
Implemented setup of coastal dataframe from BMI compliant arrays

## Additions

- Added assembly of coastal dataframe, mostly in troute_model.py
- Added script to test functionality: run_BMI_Coastal.py, located in Lower Colorado V4 example
- Made some modifications to the bmi_array2df and bmi_df2array functions, and how they are imported

## Removals

-

## Changes

-

## Testing

1. run the script run_BMI_Coastal.py, located in Lower Colorado V4 example

## Notes

-

## Todos

-

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ ] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

**Comment by JurgenZach-NOAA** [comment]
OK: fixed it, checks have passed.

File: src/troute_model.py [hunk 1]
```diff
@@ -159,6 +160,44 @@ def run(self, values: dict, until=300):
...
+
+            # Decode station ID axis
+            stationAxisName = 'station_name'
+            df_withTimes_withStations_coastal = a2df._stations_retrieve_from_arrays(\
+                            df_withTimes_coastal, self._stationArray_coastal, \
+                            self._stationStringLengthArray_coastal, stationAxisName)
+
+            self._network._coastal_boundary_depth_df = df_withTimes_withStations_coastal
+
+        #import pdb; pdb.set_trace()
```
Comment (shorvath-noaa) [inline src/troute_model.py:196]: can you remove this set_trace()?

File: test/LowerColorado_TX_v4/run_BMI_Coastal.py [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    troute.set_value('coastal_timeRef', coastal_timeRef)
+    troute.set_value('depthArray_coastal', depthArray_coastal)
+
+
+    value_dict['coastal_waterlevelArray'] = np.array([1.0, 2.0, 3.0, 4.0, 5.0], dtype=np.float64)
+    value_dict['coastal_bedlevelArray'] = np.array([12.0, 12.0, 13.0, 14.0, 15.0], dtype=np.float64)
+    value_dict['coastal_nexusID_array'] = np.array([0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5], dtype=np.int16)
+    value_dict['coastal_nexusID_StrLength_array'] = np.array([3, 3, 3, 3, 3], dtype=np.int16)
+    value_dict['coastal_timeArray'] = np.array([100000, 100000, 100000, 100000, 100000], dtype=np.int16)
+    value_dict['coastal_timeRef'] = 100
```
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/run_BMI_Coastal.py:104]: are these being used somewhere? Or was this a placeholder?

Comment (JurgenZach-NOAA) [reply test/LowerColorado_TX_v4/run_BMI_Coastal.py]: Removed the value_dict and other debugging statements.

Comment (JurgenZach-NOAA) [reply test/LowerColorado_TX_v4/run_BMI_Coastal.py]: Implemented station IDs as numpy int 64 array

File: test/LowerColorado_TX_v4/boundary_forcing/FlowFM_0000_his.nc
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/boundary_forcing/FlowFM_0000_his.nc:1]: Can you remove this file from the PR? This is just for internal testing.

File: src/troute_model.py [hunk 3]
```diff
@@ -158,7 +159,39 @@ def run(self, values: dict, until=300):
         
         self._network._qlateral = qlats_df
 
-        self._network._coastal_boundary_depth_df = pd.DataFrame(values['coastal_boundary__depth'])
+        self._network._coastal_boundary_depth_df = pd.DataFrame(values['coastal_boundary_depth'])
```
Comment (shorvath-noaa) [inline src/troute_model.py:162]: I think you can delete this line now, right?

File: test/LowerColorado_TX_v4/output/StandardAnA/202304020000.flowveldepth.nc
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/output/StandardAnA/202304020000.flowveldepth.nc:1]: Can you remove these output files?

File: test/LowerColorado_TX_v4/output/StandardAnA/202304020100.flowveldepth.nc
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/output/StandardAnA/202304020100.flowveldepth.nc:1]: Remove please

File: test/LowerColorado_TX_v4/output/StandardAnA/202304020200.flowveldepth.nc
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/output/StandardAnA/202304020200.flowveldepth.nc:1]: Remove please

File: test/LowerColorado_TX_v4/restart/channel_restart_202304020300
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/restart/channel_restart_202304020300:1]: Remove please

File: test/LowerColorado_TX_v4/restart/waterbody_restart_202304020300
Comment (shorvath-noaa) [inline test/LowerColorado_TX_v4/restart/waterbody_restart_202304020300:1]: Remove please

**Review by shorvath-noaa — CHANGES_REQUESTED** [review]

File: test/LowerColorado_TX_v4/restart/waterbody_restart_202304020300
Comment (JurgenZach-NOAA) [reply test/LowerColorado_TX_v4/restart/waterbody_restart_202304020300:1]: All files removed as requested

**Review by shorvath-noaa — APPROVED** [review]

**Pull request merged**